### DISCUSSION TITLE
[TypeScript] How to do Multiple Inheritance to create a custom element

### DISCUSSION THREAD

**Discussion by shumyvoda1215** [clientIO/joint#2086] [Q&A]
### Introduction

I need some help or suggestions on how to use multiple inheritance to create custom elements in my Angular application. Currently, I have custom Link and Element classes, but I need to introduce a common class for both. The issue is that TypeScript doesn't support multiple inheritances. I cannot use the straightforward method of inheriting from both classes for my custom elements.

Here's an example:
```ts
export class FooLink extends dia.Link {
    initialize(attributes): void {
        super.initialize(attributes);
        
        this.on('change:customFooLinkAttr', this.onCustomFooLinkAttrChange);
    }

    onCustomFooLinkAttrChange() {}
}

export class FooElement extends dia.Element {
    initialize(attributes): void {
        super.initialize(attributes);
        
        this.on('change:customFooElementAttr', this.onCustomFooElementAttrChange);
    }
    
    onCustomFooElementAttrChange() {}
}
```

And here's what I want to achieve:
```ts
export class FooCommon  {
    initialize(attributes): void {
        super.initialize(attributes);
        
        this.on('change:position', this.onPositionChange);
    }
    
    onPositionChange() {}
}

export class FooLink extends FooCommon, dia.Link {
    initialize(attributes): void {
        super.initialize(attributes);
        
        this.on('change:customFooLinkAttr', this.onCustomFooLinkAttrChange);
    }

    onCustomFooLinkAttrChange() {}
}

export class FooElement extends FooCommon, dia.Element {
    initialize(attributes): void {
        super.initialize(attributes);
        
        this.on('change:customFooElementAttr', this.onCustomFooElementAttrChange);
    }
    
    onCustomFooElementAttrChange() {}
}
```

However, the above code does not work due to TypeScript's lack of support for multiple inheritances. I am currently using TypeScript mixins to achieve what I want, but it's causing some typing problems and other issues.

Is there a proper way to implement multiple inheritance in this scenario? I would appreciate any suggestions you might have.

### Does your question relate to JointJS or JointJS+. Select both if applicable.

JointJS, JointJS+

**Comment by kumilingus** [comment]
Can't you introduce a common `interface`?

```ts
interface Common {}

class A extends dia.Element implements Common {}

class B extends dia.Link implements Common {}
```

**Reply by shumyvoda1215** [reply]
But interfaces doesn't have an actual logic. What I need is equivalent of `extend()` but in more of a TypeScript way. What I mean by `extend()` is:
```ts
var FooElement = dia.Element
    .extend(FooCommonClass)
    .extend(BarCommonClass)
    .extend({..additional logic})
    
var FooLink = dia.Link
    .extend(FooCommonClass)
    .extend(BarCommonClass)
    .extend({...additional logic})
```

**Reply by kumilingus** [reply]
The problem is that `dia.Link` and `dia.Element` both inherit from `dia.Cell`:
```ts
class dia.Link extends dia.Cell {}
class dia.Element extends dia.Cell {}
```

And you can not insert another class in between them:
```ts
class MyCell extends dia.Cell {}
class dia.Link extends MyCell {}
class dia.Element extends MyCell {}
```

Perhaps you could introduce some sort of a wrapper and use that instead of JointJS cells directly.

```ts
class CellWrapper {
   constructor(cell:  dia.Cell) { this.cell =  cell; }
   cell: dia.Cell;
   fn() {
       console.log(this.cell.id);
    }
}
class Logic1 extends CellWrapper {}
class Logic2 extends Logic1 {}
class ElementWrapper extends Logic2 { }
class LinkWrapper extends Logic2 {}

const myElement = new ElementWrapper(element1);
const myLink = new LinkWrapper(link1);
myElementWrapper.fn();
myLinkWrapper.fn();
```

**Reply by shumyvoda1215** [reply]
Thanks for your response!

> And you can not insert another class in between them.

Could you please explain a little more about why it cannot be done this way? If you want to create custom Element and Link classes with additional common logic e.g modifying position, size, and ID, you can't I have a common class in between. This is similar to what extend does, which as I understand can be considered a type of mixin.

> Perhaps you could introduce some sort of a wrapper and use that instead of JointJS cells directly.

Do you mean using a something like factory? While using a wrapper class may be a solution, if you have a lot of common classes, won't it become too complex?

I was able to implement a solution using mixins which looks something like this:
```ts
class Element1 extends Common1Mixin(Common2Mixin(Common3Mixin(dia.Element))) {}

class Element2 extends Common1Mixin(dia.Element)) {}

class Link1 extends Common1Mixin(Common2Mixin(dia.Link)) {}

function Common1Mixin<TBase extends RappidCellConstructor>(Base: TBase) {
    return class Common1 extends Base {}
}

type CellConstructor<T = dia.Cell> = new (...args: any[]) => T;

// The original Constructor type shown as an example by TypeScript is:
// type Constructor<T = {}> = new (...args: any[]) => T;
// But it knows nothing about the rappid
```

But it is so complex...

**Reply by shumyvoda1215** [reply]
Maybe I can even explore using a factory pattern. This can help simplify the code and make it more modular.

The basic idea is to create a factory function that takes a base class and a list of mixins, and returns a new class that extends the base class with the mixins applied, like:
```ts
function withMixin(base, classes) {}

class Common1 {}

class Common2 {}

class Common3 {}

class Element1 extends withMixins(dia.Element, [Common1, Common2, Common3]) {}

class Element2 extends withMixins(dia.Element, [Common1]) {}

class Link1 extends withMixins(dia.Link, [Common1, Common2]) {}

```
However, the common classes and factory used in this example implementation are not aware of Rappid, and it's unclear how to make them aware of it. 

So that is why I am here, maybe someone has same problem and can share thoughts =)
